### PR TITLE
Add MoE model (Mixtral) support. 

### PR DESCRIPTION
**Pull request by nuzant**
Changes:

- Add MoE model support (Mixtral), referece: [Megatron-LM MoE implementation](https://github.com/NVIDIA/Megatron-LM/tree/main/megatron/core/transformer/moe).  

- Add a global stats tracker to log stats that cannot be directly obtained in interface implementation (e.g. aux loss in MoE router).

- Update README.md to acknowledge open-source projects references.

### CODE REVIEW

File: realhf/api/from_hf/mixtral.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+                f"model.layers.{idx}.block_sparse_moe.experts.{j}.w2.{k}",
+                f"model.layers.{idx}.block_sparse_moe.experts.{j}.w3.{k}",
+            ]
+        if idx == config.n_layers - 1:
+            names += [f"model.norm.{k}"]
+    return names
+
+
+def get_real_config_mixtral() -> ReaLModelConfig:
+    hf_config = transformers.MixtralConfig(
```
Comment (garrett4wade) [inline realhf/api/from_hf/mixtral.py:212]: Shrink the hidden size using the global variable in `base/constants`. See other HF implementations.

File: realhf/impl/model/interface/rw_interface.py [hunk 2]
```diff
@@ -166,6 +166,9 @@ def train_step(
         )
 
         res = {}
+        global_stats = constants.log_global_stats_tracker(
+            return_dict=True, clear_stats_after_logging=True
+        )
```
Comment (garrett4wade) [inline realhf/impl/model/interface/rw_interface.py:171]: Log in the evaluate method?

File: realhf/impl/model/modules/attn.py [hunk 3]
```diff
@@ -86,6 +87,7 @@ def __init__(
                 n_q_heads * head_dim,
                 hidden_dim,
                 bias=use_attn_proj_bias,
+                sequence_parallel=constants.sequence_parallel(),
```
Comment (garrett4wade) [inline realhf/impl/model/modules/attn.py:90]: This is dangerous because the `sequence_parallel` variable may not be set when importing this file. Let's remove this parameter and forcely set `sequence_parallle=constants.sequence_parallel` in the `__init__` method.

File: realhf/impl/model/modules/mlp.py [hunk 4]
```diff
@@ -306,42 +307,52 @@ def __init__(
...
+        is_expert: bool = False,
         # dtype and device
         dtype: Optional[torch.dtype] = None,
         device: Optional[Union[str, torch.device]] = None,
     ):
         super().__init__()
-        sequence_parallel = constants.sequence_parallel()
-        if not model_parallel and (sequence_parallel or gradient_accumulation_fusion):
+        # when used as experts the MLP always compute without sequence parallel
+        self.sequence_parallel = constants.sequence_parallel() and not is_expert
```
Comment (garrett4wade) [inline realhf/impl/model/modules/mlp.py:324]: Why not combine sequence parallel and MoE?

File: realhf/impl/model/nn/real_llm_base.py [hunk 5]
```diff
@@ -482,6 +496,17 @@ def real_model_tblock_param_keys(
...
+        num_experts = config.moe.num_experts
+        keys += [
+            f"{idx + 1}.mlp.router.weight",
+        ]
+        for j in range(num_experts):
+            keys += [
+                f"{idx + 1}.mlp.experts.local_experts.{j}.gate_proj.weight",
+                f"{idx + 1}.mlp.experts.local_experts.{j}.up_proj.weight",
+                f"{idx + 1}.mlp.experts.local_experts.{j}.down_proj.weight",
+            ]
```
Comment (garrett4wade) [inline realhf/impl/model/nn/real_llm_base.py:509]: Does the key have the same order as in the state dict? I think the actual order should be

```python
for j in range(num_experts):
    keys += [f"{idx+1}.router", f"{idx+1}.w1", f"{idx+1}.w2", f"{idx+1}.w3"]
```

If the order is not the same, it will be a bug in parameter reallocation.

File: realhf/impl/model/parallelism/model_parallel/modules.py [hunk 6]
```diff
@@ -919,6 +912,7 @@ def __init__(
         output_size,
         bias=True,
         input_is_parallel=True,
+        sequence_parallel=False,
```
Comment (garrett4wade) [inline realhf/impl/model/parallelism/model_parallel/modules.py:915]: Why as an argument instead of a runtime option, e.g., `constants.sequence_parallel()`? We may want to enable sequence parallel during training but disable it during inference or generation.

File: realhf/impl/model/parallelism/model_parallel/modules.py [hunk 7]
```diff
@@ -1014,21 +1008,16 @@ def parallel_lm_logits(
     input_: torch.HalfTensor,
     word_embeddings_weight: torch.HalfTensor,
     parallel_output: bool = False,
-    async_tensor_model_parallel_allreduce: bool = False,
+    sequence_parallel: bool = False,
```
Comment (garrett4wade) [inline realhf/impl/model/parallelism/model_parallel/modules.py:1011]: .

File: realhf/impl/model/utils/grouped_gemm.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2023, NVIDIA CORPORATION. All rights reserved.
+
+try:
+    import grouped_gemm
+except ImportError:
+    grouped_gemm = None
+
+
+def grouped_gemm_is_available():
```
Comment (garrett4wade) [inline realhf/impl/model/utils/grouped_gemm.py:9]: Move it into constants?

File: realhf/impl/model/utils/moe.py [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+            exceed_mask = torch.gather(drop_mask, 1, top_indices)
+            final_probs = probs * torch.logical_not(exceed_mask)
+            final_indices = top_indices.clone().masked_fill_(
+                exceed_mask, torch.iinfo(torch.long).max
+            )
+            tokens_per_expert_before_capacity = topk_mask.sum(dim=0)
+        return final_probs, final_indices, tokens_per_expert_before_capacity
+
+
+class moe_gather(torch.autograd.Function):
```
Comment (garrett4wade) [inline realhf/impl/model/utils/moe.py:391]: We still require this function even without expert parallel, don't we?

**Review by garrett4wade — COMMENTED** [review]
LGTM. Several minor points should be further addressed.

File: realhf/impl/model/utils/moe.py (hunk 9 above)
Comment (nuzant) [reply realhf/impl/model/utils/moe.py]: removed, this is not needed

File: realhf/impl/model/nn/real_llm_base.py (hunk 5 above)
Comment (nuzant) [reply realhf/impl/model/nn/real_llm_base.py]: Fixed

Comment (garrett4wade) [reply realhf/impl/model/nn/real_llm_base.py]: My previous comment was not correct. The order has been checked to be coorect, right?

File: realhf/impl/model/modules/mlp.py (hunk 4 above)
Comment (nuzant) [reply realhf/impl/model/modules/mlp.py]: This is because the gather and scatter of sequence parallel are done in MoE token dispatcher before and after permutation. I did not re-scatter them again before feeding permuted hidden states into experts because it seems meaningless and only add communication overheads.  Therefore even when sequence parallel is enabled, expert MLPs behave as sequence parallel is not enabled. 

Maybe there is a clever way to permute hidden states with sequence parallel to save memory and reduce communication overhead but  I did not think of one.

File: realhf/impl/model/modules/attn.py (hunk 3 above)
Comment (nuzant) [reply realhf/impl/model/modules/attn.py]: Fixed.

File: realhf/impl/model/interface/rw_interface.py (hunk 2 above)
Comment (nuzant) [reply realhf/impl/model/interface/rw_interface.py:171]: Fixed

File: realhf/api/from_hf/mixtral.py (hunk 1 above)
Comment (nuzant) [reply realhf/api/from_hf/mixtral.py:221]: Fixed

File: realhf/impl/model/utils/grouped_gemm.py (hunk 8 above)
Comment (nuzant) [reply realhf/impl/model/utils/grouped_gemm.py:9]: I think they should be moved into experts.py since they are not used anywhere else.

File: realhf/impl/model/modules/attn.py [hunk 10]
```diff
@@ -4,6 +4,7 @@
 import torch.nn as nn
 import torch.utils.checkpoint
 
+import realhf.base.constants as constants
```
Comment (garrett4wade) [inline realhf/impl/model/modules/attn.py:7]: Remove this line.

File: realhf/impl/model/nn/real_llm_base.py (hunk 5 above)
Comment (nuzant) [reply realhf/impl/model/nn/real_llm_base.py:507]: `['1.attn.c_attn.ln.weight', '1.attn.c_attn.q_attn.weight', '1.attn.c_attn.k_attn.weight', '1.attn.c_attn.v_attn.weight', '1.attn.c_proj.weight', '1.mlp.ln.weight', '1.mlp.router.weight', '1.mlp.experts.local_experts.0.gate_proj.weight', '1.mlp.experts.local_experts.0.up_proj.weight', '1.mlp.experts.local_experts.0.down_proj.weight', '1.mlp.experts.local_experts.1.gate_proj.weight', '1.mlp.experts.local_experts.1.up_proj.weight', '1.mlp.experts.local_experts.1.down_proj.weight', '1.mlp.experts.local_experts.2.gate_proj.weight', '1.mlp.experts.local_experts.2.up_proj.weight', '1.mlp.experts.local_experts.2.down_proj.weight', '1.mlp.experts.local_experts.3.gate_proj.weight', '1.mlp.experts.local_experts.3.up_proj.weight', '1.mlp.experts.local_experts.3.down_proj.weight']`

I double checked, the original one should be correct. Reverted.

File: realhf/impl/model/modules/mlp.py [hunk 11]
```diff
@@ -306,42 +303,51 @@ def __init__(
...
+        sequence_parallel = constants.sequence_parallel() and not is_expert
         if not model_parallel and (sequence_parallel or gradient_accumulation_fusion):
             global SEQUENCE_PARALLEL_WARNED
             if not SEQUENCE_PARALLEL_WARNED:
                 logger.warning(
                     "sequence_parallel and gradient_accumulation_fusion are only available in model parallel mode"
                 )
                 SEQUENCE_PARALLEL_WARNED = True
-            sequence_parallel = False
+            self.sequence_parallel = False
```
Comment (garrett4wade) [inline realhf/impl/model/modules/mlp.py:329]: Remove `self.sequence_parallel`.

File: realhf/impl/model/modules/moe/experts.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+        config: ReaLModelConfig,
+        init_method: Callable = init.xavier_normal_,
+        dtype: Optional[torch.dtype] = None,
+        device: Optional[Union[str, torch.device]] = None,
+    ):
+        super().__init__()
+        assert (
+            not constants.sequence_parallel()
+        ), "Grouped GEMM does not support sequence parallel"
+        assert dtype == torch.bfloat16, "Grouped GEMM only supports bfloat16 dtype"
```
Comment (garrett4wade) [inline realhf/impl/model/modules/moe/experts.py:118]: Input dtype None will create fp32 weights, but this assersion will not work with dtype == None.

File: realhf/impl/model/modules/moe/layer.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+            layer_norm_fn = LlamaRMSNorm
+        elif config.layer_norm_type == "gemma":
+            layer_norm_fn = GemmaRMSNorm
+        self.ln = layer_norm_fn(
+            config.hidden_dim, eps=config.layer_norm_epsilon, dtype=dtype, device=device
+        )
+
+        self.router = TopKRouter(config=self.config, layer_idx=layer_idx)
+        self.token_dispatcher = MoETokenDispatcher(config=self.config)
+        if config.moe.use_grouped_gemm:
```
Comment (garrett4wade) [inline realhf/impl/model/modules/moe/layer.py:47]: Should we use a if condition about the dtype here, or use the assersion inside the GroupedMLP class? I think the former solution is better because it does not kill the experiment.

**Review by garrett4wade — APPROVED** [review]

File: realhf/impl/model/modules/moe/layer.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+            self.experts = SequentialMLP(self.config, dtype=dtype, device=device)
+
+    def forward(self, hidden_states: torch.Tensor):
+        if (
+            self.training
+            and constants.model_parallel_world_size() > 1
+            and not constants.sequence_parallel()
+            and not self.config.moe.use_grouped_gemm
+        ):
+            logger.warning(
```
[the remark this answers is not in the document]
Comment (nuzant) [reply realhf/impl/model/modules/moe/layer.py]: This warnning is not true in this implementation, removed.

File: realhf/impl/model/parallelism/model_parallel/modules.py [hunk 15]
```diff
@@ -843,13 +836,12 @@ def forward(self, input_) -> torch.Tensor:
             - bias
         """
         bias = self.bias if not self.skip_bias_add else None
-        if self.async_tensor_model_parallel_allreduce and constants.sequence_parallel():
-            raise RuntimeError(
-                "`async_tensor_model_parallel_allreduce` and `sequence_parallel` "
-                "cannot be enabled at the same time."
-            )
+        sequence_parallel = constants.sequence_parallel() and not self.is_expert
```
[the remark this answers is not in the document]
Comment (nuzant) [reply realhf/impl/model/parallelism/model_parallel/modules.py:843]: Done

File: realhf/impl/model/modules/moe/layer.py (hunk 13 above)
Comment (nuzant) [reply realhf/impl/model/modules/moe/layer.py]: Removed assertion in GroupedMLP and added condition and warning here.

File: realhf/impl/model/modules/moe/experts.py (hunk 12 above)
Comment (nuzant) [reply realhf/impl/model/modules/moe/experts.py]: I did not quite understand the sentence, but this assertion is removed so maybe it does not matter

File: realhf/impl/model/modules/mlp.py (hunk 11 above)
Comment (nuzant) [reply realhf/impl/model/modules/mlp.py]: Done

File: realhf/impl/model/modules/moe/experts.py (hunk 12 above)
Comment (garrett4wade) [reply realhf/impl/model/modules/moe/experts.py]: My bad. I don't have a clear mind today.

**Review by garrett4wade — APPROVED** [review]
LGTM. Merged.
